script: Fix a crash for struct type handling in script

honggyukim opened the pull request:
We didn't handle struct type for scripting, so it crashes as follows.
```
  $ g++ -pg -g -O2 tests/s-dwarf7.cpp
  $ uftrace record -a --no-libcall a.out
  $ uftrace script --no-pager -S scripts/dump.py
        ...
  4981.341259000   4788: [entry] main(7fd7928006cd) depth: 0
  WARN: invalid argument format: 10
  WARN: invalid argument format: 10
  4981.341264100   4788: [entry] compare_iters(7fd79280088a) depth: 1
  Segmentation fault (core dumped)
```
This patch fixes this problem for both python and luajit, but print the
struct type name in string type as follows.
```
  $ uftrace script -S scripts/dump.py
        ...
  4981.341259000   4788: [entry] main(7fd7928006cd) depth: 0
  4981.341264100   4788: [entry] compare_iters(7fd79280088a) depth: 1
    args[0] <type 'str'>: __normal_iterator
    args[1] <type 'str'>: __normal_iterator
  4981.342013200   4788: [exit ] compare_iters(7fd79280088a) depth: 1
    retval  <type 'long'>: 0
  4981.342024300   4788: [exit ] main(7fd7928006cd) depth: 0
    retval  <type 'long'>: 0
```
Signed-off-by: Honggyu Kim <honggyu.kp@gmail.com>

namhyung commented on utils/script-luajit.c:
```diff
@@ -197,6 +197,16 @@ static void setup_argument_context(bool is_retval, struct script_context *sc_ctx
 			break;
 
+		case ARG_FMT_STRUCT:
+			if (spec->type_name) {
```
What if it doesn't have a type name?

namhyung commented on utils/script-luajit.c:
```diff
@@ -197,6 +197,16 @@ static void setup_argument_context(bool is_retval, struct script_context *sc_ctx
+		case ARG_FMT_STRUCT:
+			if (spec->type_name) {
+				slen = strlen(spec->type_name);
+				dllua_pushinteger(L, 1 + count++);
```
`++count` instead?

namhyung reviewed (commented):
I think we should give a hint to the scripts as it's just not a string.  Maybe we can add a prefix like "struct: TYPENAME" or add surrounding braces like "{ TYPENAME }".

honggyukim commented:
@namhyung Thanks for the review. I changed the output as follows.
```
$ g++ -pg -g -o t-struct tests/s-struct.c

$ uftrace script -S scripts/dump.py --no-libcall --no-event --record t-struct
uftrace_begin(ctx)
  record  : False
  version : v0.10-17-g8d1b ( x86_64 dwarf python luajit tui perf sched dynamic )
  cmds    : t-struct

11336060.26574470  26208: [entry] main(4006ef) depth: 0
11336060.26575270  26208: [entry] foo(40068f) depth: 1
  args[0] <class 'str'>: struct: Option{}
  args[1] <class 'str'>: struct: StringRef{}
  args[2] <class 'int'>: 44
  args[3] <class 'int'>: 55
  args[4] <class 'int'>: 66
11336060.26780955  26208: [exit ] foo(40068f) depth: 1
11336060.26782239  26208: [exit ] main(4006ef) depth: 0
  retval  <class 'int'>: 0

uftrace_end()
```
I also added 3 more tests for struct argument handling.

namhyung commented on tests/s-struct.c:
```diff
@@ -0,0 +1,34 @@
+        g_value2 = Value2;
+}
+
+int main()
```
C requires void in the signature.

honggyukim replied:
done

namhyung commented on tests/s-struct.c:
```diff
@@ -0,0 +1,34 @@
+        struct StringRef S;
+
+        foo(Opt, S, 44, 55, 66);
+}
```
Please add 'return 0'

honggyukim replied:
I changed this example from C++ to C, so didn't add it because C++ doesn't require to have explicit `return 0` for main function. But I changed this to C again, so you're right. I just added it :)

namhyung commented on tests/t259_arg_struct_script.py:
```diff
@@ -0,0 +1,50 @@
+
+    def build(self, name, cflags='', ldflags=''):
+        if not 'dwarf' in self.feature:
+            return TestBase.TEST_SKIP
```
I think it should also skip if 'python' is not in the feature.

honggyukim replied:
done

namhyung commented on tests/t260_arg_struct_luajit.py:
```diff
@@ -0,0 +1,45 @@
+
+    def build(self, name, cflags='', ldflags=''):
+        if not 'dwarf' in self.feature:
+            return TestBase.TEST_SKIP
```
Ditto for 'luajit'.

honggyukim replied:
done

honggyukim reviewed (commented):
Thanks for the review. I just updated again.

namhyung reviewed (approved):
LGTM
